common/elasticsearch/client/v6/client.go

```diff
@@ -0,0 +1,186 @@
+	return elastic.IsNotFound(err)
+}
+
+// NewV6Client returns a new implementation of GenericClient
+func NewV6Client(
+	connectConfig *config.ElasticSearchConfig,
+	logger log.Logger,
+	tlsClient *http.Client,
+	awsSigningClient *http.Client,
+) (*ElasticV6, error) {
+	clientOptFuncs := []elastic.ClientOptionFunc{
+		elastic.SetURL(connectConfig.URL.String()),
+		elastic.SetRetrier(elastic.NewBackoffRetrier(elastic.NewExponentialBackoff(128*time.Millisecond, 513*time.Millisecond))),
+		elastic.SetDecoder(&elastic.NumberDecoder{}), // critical to ensure decode of int64 won't lose precise)
+	}
+	if connectConfig.DisableSniff {
+		clientOptFuncs = append(clientOptFuncs, elastic.SetSniff(false))
+	}
+	if connectConfig.DisableHealthCheck {
+		clientOptFuncs = append(clientOptFuncs, elastic.SetHealthcheck(false))
+	}
+
+	if awsSigningClient != nil {
+		clientOptFuncs = append(clientOptFuncs, elastic.SetHttpClient(awsSigningClient))
+	}
+
+	if tlsClient != nil {
+		clientOptFuncs = append(clientOptFuncs, elastic.SetHttpClient(tlsClient))
+	}
+
+	client, err := elastic.NewClient(clientOptFuncs...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ElasticV6{
+		client: client,
+		logger: logger,
+	}, nil
+}
+
+func (c *ElasticV6) PutMapping(ctx context.Context, index, body string) error {
+	_, err := c.client.PutMapping().Index(index).Type("_doc").BodyString(body).Do(ctx)
+	return err
+}
+
+func (c *ElasticV6) CreateIndex(ctx context.Context, index string) error {
+	_, err := c.client.CreateIndex(index).Do(ctx)
+	return err
+}
+
+func (c *ElasticV6) Count(ctx context.Context, index, query string) (int64, error) {
+	return c.client.Count(index).BodyString(query).Do(ctx)
+}
+
+func (c *ElasticV6) ClearScroll(ctx context.Context, scrollID string) error {
+	return elastic.NewScrollService(c.client).ScrollId(scrollID).Clear(ctx)
+}
+func (c *ElasticV6) Scroll(ctx context.Context, index, body, scrollID string) (*client.Response, error) {
+	scrollService := elastic.NewScrollService(c.client)
+	var esResult *elastic.SearchResult
+	var err error
+
+	// we are not returning error immediately here, as result + error combination is possible
+	if len(scrollID) == 0 {
+		esResult, err = scrollService.Index(index).Body(body).Do(ctx)
+	} else {
+		esResult, err = scrollService.ScrollId(scrollID).Do(ctx)
+	}
+
+	if esResult == nil {
+		return nil, err
+	}
+
+	var hits []*client.SearchHit
+	if esResult.Hits != nil {
+		for _, h := range esResult.Hits.Hits {
+			hits = append(hits, &client.SearchHit{Source: *h.Source})
+		}
+	}
+
+	result := &client.Response{
+		TookInMillis: esResult.TookInMillis,
+		TotalHits:    esResult.TotalHits(),
+		Hits:         &client.SearchHits{Hits: hits},
+		ScrollID:     esResult.ScrollId,
+	}
+
+	if len(esResult.Aggregations) > 0 {
+		result.Aggregations = make(map[string]json.RawMessage, len(esResult.Aggregations))
+		for key, agg := range esResult.Aggregations {
+			result.Aggregations[key] = *agg
```

Comment:
it is not safe dereferencing a pointer without checking if it's nil

Reply:
done